src/commands/classified/external.rs

```diff
@@ -70,112 +170,62 @@ async fn run_with_iterator_arg(
+            let home_dir = dirs::home_dir();
+            let path = &path;
+            let args = command.args.clone();
+
+            let it_replacement = match nu_value_to_string(&command, &value) {
+                Ok(value) => value,
+                Err(reason) => {
+                    yield Ok(Value {
+                        value: UntaggedValue::Error(reason),
+                        tag: name_tag
+                    });
+                    return;
+                }
+            };
 
-    let input_strings = inputs
-        .iter()
-        .map(|i| match i {
-            Value {
-                value: UntaggedValue::Primitive(Primitive::Int(i)),
-                ..
-            } => Ok(i.to_string()),
-            Value {
-                value: UntaggedValue::Primitive(Primitive::String(s)),
-                ..
-            }
-            | Value {
-                value: UntaggedValue::Primitive(Primitive::Line(s)),
-                ..
-            } => Ok(s.clone()),
-            Value {
-                value: UntaggedValue::Primitive(Primitive::Path(p)),
-                ..
-            } => Ok(p.to_string_lossy().to_string()),
-            _ => {
-                let arg = args.iter().find(|arg| arg.contains("$it"));
-                if let Some(arg) = arg {
-                    Err(ShellError::labeled_error(
-                        "External $it needs string data",
-                        "given row instead of string data",
-                        &arg.tag,
-                    ))
+            let process_args = args.iter().filter_map(|arg| {
+                if arg.chars().all(|c| c.is_whitespace()) {
+                    None
                 } else {
-                    Err(ShellError::labeled_error(
-                        "$it needs string data",
-                        "given something else",
-                        &name_tag,
-                    ))
+                    let arg = expand_tilde(arg.deref(), || home_dir.as_ref());
+                    Some(arg.replace("$it", &it_replacement))
                 }
-            }
-        })
-        .collect::<Result<Vec<String>, ShellError>>()?;
+            }).collect::<Vec<_>>();
 
-    let home_dir = dirs::home_dir();
-    let commands = input_strings.iter().map(|i| {
-        let args = args.iter().filter_map(|arg| {
-            if arg.chars().all(|c| c.is_whitespace()) {
-                None
-            } else {
-                let arg = shellexpand::tilde_with_context(arg.deref(), || home_dir.as_ref());
-                Some(arg.replace("$it", &i))
+            match spawn(&command, &path, &process_args, None, is_last).await {
+                Ok(res) => {
+                    if let Some(mut res) = res {
+                        while let Some(item) = res.next().await {
+                            yield Ok(item)
+                        }
+                    }
+                }
+                Err(reason) => {
+                    yield Ok(Value {
```

Comment:
Wouldn't you just yield out an Err instead of an Ok here?  And then immediately return

Reply:
The functions returns an optional `InputStream`. If we yield an `error` instead, Nu will panic since `.to_input_stream()` doesn't allow Err's.

Reply:
Oh oh, got it.  I should read more closely :p